Add org.sil.scripture-app-builder

chrisvire opened the pull request:
<!-- ⚠️  The submission PR must be against the `new-pr` branch ⚠️  -->

### Please confirm your submission meets all the criteria

<!-- Please replace each `[ ]` with `[X]` when the step is complete -->

- [x] Please describe the application briefly.
Create customized mobile apps with synchronized Scripture text and audio

- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:**


<!-- ⚠️  Don't modify anything below this line ⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

chrisvire commented:
We have 4 related but separate "App Builders" that we are switching from Debian packages for Ubuntu to Flatpak.

Scripture App Builder (this one) 
- home: https://software.sil.org/scriptureappbuilder/
- repo: https://github.com/sillsdev/flathub/tree/org.sil.scripture-app-builder
Dictionary App Builder
- home: https://software.sil.org/dictionaryappbuilder/
- repo: https://github.com/sillsdev/flathub/tree/org.sil.dictionary-app-builder
Reading App Builder
- home: https://software.sil.org/readingappbuilder/
- repo: https://github.com/sillsdev/flathub/tree/org.sil.reading-app-builder
Keyboard App Builder
- home: https://software.sil.org/keyboardappbuilder/
- repo: https://github.com/sillsdev/flathub/tree/org.sil.keyboard-app-builder

Should we submit them all at once or work through any issues with this one first and then submit the other 3 later (and make the appropriate fixes to them if anything is found in this one)?

Thanks!
Chris

bbhtt commented:
How large are these and are they related to one another ? If size is not an issue having them in one Flatpak is better similar to how Libreoffice works.

bbhtt commented on org.sil.scripture-app-builder.desktop:
```diff
@@ -0,0 +1,11 @@
+Version=1.0
+Name=Scripture App Builder
+Comment=Build customized Scripture apps for Android smartphones and tablets
+Exec=/app/bin/sab-run.sh %u
```
```suggestion
Exec=sab-run.sh %u
```

chrisvire replied:
Moved to source repo and fixed.

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+  # All external devices (catch all for devices we interact with and graphical acceleration)
+  - --device=all
+  # Local file browsing access
+  - --filesystem=host
```
why is this needed?

chrisvire replied:
The application builds mobile apps based on project files that are in the file system. It also needs to browse to add audio files and text files to the project from different locations in the file system.

After a build completes, it uses adb to install the app to an Android device so it needs to connect to the device through USB.

chrisvire replied:
The application builds android apps from projects located on the user's computer.  Once the android app is built, it connects via usb to the phone to install and run the app.

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+  - --filesystem=host
+  # Temporary file access
+  - --filesystem=/tmp
+  # Filesystem mounts
+  - --filesystem=/mnt
```
```suggestion
```

host includes this

chrisvire replied:
Removed

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+  # Local file browsing access
+  - --filesystem=host
+  # Temporary file access
+  - --filesystem=/tmp
```
why is /tmp needed?

chrisvire replied:
The build process writes files to /tmp.

bbhtt replied:
These permissions have no effect on the build process.

chrisvire replied:
Sorry, when I said the build process, I should have said the mobile app build process which is done during the running of the desktop app.

bbhtt replied:
> Sorry, when I said the build process, I should have said the mobile app build process which is done during the running of the desktop app.

Does it need to be `/tmp` from host? You can set the TMPDIR variable to XDG_RUNTIME_DIR or `/var/tmp`. One is tmpfs, the second one is persistent.

chrisvire replied:
I did some testing.  Access to /tmp was being used for builds.  I changed the sab-run.sh to have `-Djava.io.tmpdir=/var/tmp` so that the default tmp dir is set to /var/tmp instead of /tmp.

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+  # USB Drives
+  - --system-talk-name=org.freedesktop.Udisks2
+  # File saving access
+  - --system-talk-name=org.freedesktop.FileManager1
```
this does not run on system bus.
```suggestion
  - --talk-name=org.freedesktop.FileManager1
```

chrisvire replied:
Changed

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+  # File saving access
+  - --system-talk-name=org.freedesktop.FileManager1
+
+# Build args give resource access during the build
+build-options:
+  build-args:
+    - --share=network
+
```
Network access during build is not allowed on Flathub

chrisvire replied:
We download a docker image of the pre-built binaries from a private repo. Is this flag not necessary to do that?

bbhtt replied:
This is clearly mentioned in the requirements page linked in the PR checklist.

You will have to find an alternate building mechanism

bbhtt replied:
If you want to use prebuilt binaries, you can host them somewhere and directly use those in the source. Not the docker image.

chrisvire replied:
We will be moving the prebuilt binaries to an S3 Bucket and access through url and sha256

chrisvire replied:
Removed build-args for `--share=network` and using zip file from S3 bucket.

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+- name: scripture-app-builder
+  buildsystem: simple
+  build-commands:
+    - pwd
+    # Source tarball from a docker image and extract to flatpak app
+    - cd /tmp
```
```suggestion
```

chrisvire replied:
What is different?

chrisvire replied:
Removed

bbhtt commented on python3-aeneas.yml:
```diff
@@ -0,0 +1,99 @@
+  # was 1.22.4: https://pypi.org/project/numpy/1.22.4/#files
+name: python3-aeneas
+buildsystem: simple
+build-options:
+    append-ld-library-path: /app/lib
```
```suggestion
```

/app/lib is already in ld path

chrisvire replied:
Removed

bbhtt commented on python3-aeneas.yml:
```diff
@@ -0,0 +1,99 @@
+build-options:
+    append-ld-library-path: /app/lib
+build-commands:
+    - echo PWD="${PWD}"  and  FLATPAK_DEST="${FLATPAK_DEST}"
```
```suggestion
```

chrisvire replied:
What is different? Replace all 4 lines with just the 1 line?

This file comes from a app already accepted into flathub:

https://github.com/flathub/org.sil.Bloom/blob/master/python3-aeneas.yml

bbhtt replied:
$PWD is a shell built in and $FLATPAK_DEST is already set. The source you are pointing at is doing it wrong.

echo -ing this is not a necessary part of the build process

chrisvire replied:
Removed

bbhtt commented on python3-aeneas.yml:
```diff
@@ -0,0 +1,99 @@
+    append-ld-library-path: /app/lib
+build-commands:
+    - echo PWD="${PWD}"  and  FLATPAK_DEST="${FLATPAK_DEST}"
+    - python3 -m pip install --upgrade pip
```
```suggestion
```

chrisvire replied:
Removed

bbhtt commented on python3-aeneas.yml:
```diff
@@ -0,0 +1,99 @@
+build-commands:
+    - echo PWD="${PWD}"  and  FLATPAK_DEST="${FLATPAK_DEST}"
+    - python3 -m pip install --upgrade pip
+    # Ensuring python3 points to the right place if needed
+    #- rm -f /app/bin/python
+    - python3 -m pip list
```
```suggestion
```

chrisvire replied:
What is different?

bbhtt replied:
not a part of any build process.

chrisvire replied:
Removed

bbhtt commented on python3-aeneas.yml:
```diff
@@ -0,0 +1,99 @@
+      url: https://files.pythonhosted.org/packages/4c/f3/038b302fdfbe3be7da016777069f26ceefe11a681055ea1f7817546508e3/soupsieve-2.5-py3-none-any.whl
+      sha256: eaa337ff55a1579b6549dc679565eac1e3d000563bcb1c8ab0d0fefbc0c2cdc7
+    
+    #OLD
```
Can you remove the commented out sources?

chrisvire replied:
Removed

bbhtt commented on python3-aeneas.yml:
```diff
@@ -0,0 +1,99 @@
+    #   sha256: 58d5c3d29f5a36ffeb94f02f0d786cd53014cf9b3b3951d42e0080d8a9498d30
+modules:
+    - name: espeak
+      buildsystem: simple
+      build-commands:
+          - |
+              set -xueo pipefail
+              cd src
+              make install
```
Use `no-autogen: true`, `subdir: src`, drop build-commands and buildsystem: simple

chrisvire replied:
Changed to your recommendation

bbhtt commented on python3-requests.yaml:
```diff
@@ -0,0 +1,23 @@
+# Generated with flatpak-pip-generator --runtime=org.freedesktop.Sdk//23.08 requests --yaml
+name: python3-requests
+buildsystem: simple
+build-options:
+    append-ld-library-path: /app/lib
```
```suggestion
```

chrisvire replied:
Removed

bbhtt commented on sab-node20.yml:
why not just use the node 20 extension?

chrisvire replied:
The node 20 extension was building node from source with yarn and taking an extra 20 minutes for a build.  This extract a pre-built node 20.

bbhtt replied:
> The node 20 extension was building node from source with yarn and taking an extra 20 minutes for a build. This extract a pre-built node 20.

Do you need node available at runtime for this?

If you use the node20 extension there is no need to rebuild node itself. You can copy the contents of that extension into `/app`

It provides two scripts to copy the contents `install.sh, install-sdk.sh`

chrisvire replied:
We need the full sdk at runtime since we do a build of a Progressive Web App (PWA) from the user's project data + a PWA template (https://github.com/sillsdev/appbuilder-pwa).

Our application is expecting it to be installed in a specific place (and be a specific version) since we can't just use whatever node is on the path (we have had problems in the past).

If we used the extension, I would need to make code changes to the source of the app to look for it in the install location.

Is it ok to do this?

chrisvire replied:
I have switched to using the node20 extension.  I assume this is OK:

`ln -s /app/node /app/scripture-app-builder/tools/node`

bbhtt replied:
> I have switched to using the node20 extension. I assume this is OK:

Yes if it work.

chrisvire replied:
It works for me.  Thanks

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+    - pwd
+    # Source tarball from a docker image and extract to flatpak app
+    - cd /tmp
+    - python3 /run/build/scripture-app-builder/docker_pull.py ghcr.io/sillsdev/scripture-app-builder:12.1
+    - python3 /run/build/scripture-app-builder/extract_layer.py /app/output
```
why pull from a docker image? Where is the source of the application ?

bbhtt replied:
Fwiw this will not work on Flathub, there is no network access during build.

chrisvire replied:
Due to relationships with Bible societies, the source code is in a private repo.  We do a Build in GitHub Actions which creates a tarball of the artifacts and imports that into a docker image. Since we have a nonprofit GitHub account, there is no cost for storing this in GHCR.  It is an economical way to store the results of a build to share and is versioned with tags.

[Saving Images and Containers as Tar Files for Sharing](https://dockerlabs.collabnix.com/beginners/saving-images-as-tar/)

Here is the step we use in our GitHub Action:
```
      - name: "Push AppBuilders Tarball to GHCR"
        run: |
          docker import release/output/app-builders.tgz app-builders:latest
          docker tag app-builders:latest ghcr.io/$OWNER/app-builders:latest
          docker tag app-builders:latest ghcr.io/$OWNER/app-builders:${{ steps.version.outputs.SAB }}
          docker push ghcr.io/$OWNER/app-builders:latest
          docker push ghcr.io/$OWNER/app-builders:${{ steps.version.outputs.SAB }}
```

I don't quite understand `there is no network access during build`. How would installing python modules or OpenJFX into the Flatpak image work without network access?

chrisvire replied:
So we can't specify a docker image to download and then extract the files?  It needs to be a zip file hosted somewhere with a sha256?

hfiguiere replied:
docker images are not an archive format.

chrisvire replied:
Will change to tar ball stored in an S3 bucket.

chrisvire replied:
Removed and using zip from S3 bucket.

bbhtt commented on flathub.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "only-arches": ["x86_64"]
+}
```
If the end result is x86_64 only why are there aarch64 sources around ? Like OpenJFX

chrisvire replied:
Actually, I just noticed that the OpenJFX with aarch64 is faulty and should be removed.  We are using Java 17 which is required due to some built libraries being shared between the desktop and android compiled app and Java 17 is the correct version to use with the Android Plugin for Gradle.  There are no builds of OpenJFX for Java 17 and aarch64.

chrisvire replied:
Removed incorrect aarch64 reference to just documenation.

bbhtt commented on docker_pull.py:
You can use it directly from the source instead of putting it here.

```yaml
- type: file
  url: https://raw.githubusercontent.com/NotGlop/docker-drag/2d9880e2e1756d3bc8f265e8a30a662387ab67ad/docker_pull.py
  sha256: 
```

chrisvire replied:
We were worried about it going away and breaking our build.  It is pretty essential for the build to work.

chrisvire replied:
Looking at the license, it is GPL3, so we should pull it from their repo to use it since our license is MIT?

chrisvire replied:
Removed.

bbhtt commented:
Desktop file and metainfo file belongs upstream.

hfiguiere commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+    - python3 /run/build/scripture-app-builder/docker_pull.py ghcr.io/sillsdev/scripture-app-builder:12.1
+    - python3 /run/build/scripture-app-builder/extract_layer.py /app/output
+    - rm sillsdev_scripture-app-builder.tar
+    - cd /run/build/scripture-app-builder
```
this should already be the current directory. If you get out of part of these steps then it's too complicated.

chrisvire replied:
When I switch from docker file to tarball, this should go away.

chrisvire replied:
Removed

hfiguiere commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+      #- mkdir /app/lib
+    - mv sdk /app/lib/sdk
+    # Icons, desktop entry, metainfo and runner script
+    - mkdir -p /app/share/icons/hicolor/128x128/apps
+    - mkdir -p /app/share/icons/hicolor/scalable/apps
+    - mv /app/output/app-icons/scripture-app-builder.png /app/share/icons/hicolor/128x128/apps/org.sil.scripture-app-builder.png
+    - mv /app/output/app-icons/scripture-app-builder.svg /app/share/icons/hicolor/scalable/apps/org.sil.scripture-app-builder.svg
```
use `install`. It creates the target directories at the same tieme.

chrisvire replied:
Changed to using `install`.

hfiguiere commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,114 @@
+    - desktop-file-install --dir /app/share/applications org.sil.scripture-app-builder.desktop
+    - install -Dm644 org.sil.scripture-app-builder.metainfo.xml -t /app/share/metainfo/
+    - install -Dm755 sab-run.sh /app/bin/sab-run.sh
+    - rm -rf /app/output/*
```
maybe output shouldn't be in `/app`

chrisvire replied:
Stopped using /app/output

hfiguiere commented on python3-aeneas.yml:
this should be in the main manifest since it's not autogenerated.

chrisvire replied:
The definition is "shared" with [org.sil.Bloom](https://github.com/flathub/org.sil.Bloom/) and easier to maintain.

chrisvire replied:
I have combined python3-aeneas.xml into the main manifest.

hfiguiere commented on sab-run.sh:
```diff
@@ -0,0 +1,16 @@
+# echo " "
+# echo "$@"
+# echo " "
+exec /app/jre/bin/java --module-path "/app/lib/sdk" --add-modules javafx.web,javafx.fxml,javafx.swing,javafx.media --add-opens=javafx.fxml/javafx.fxml=ALL-UNNAMED -jar /app/scripture-app-builder/bin/scripture-app-builder.jar $@
```
```suggestion
exec /app/jre/bin/java --module-path "/app/lib/sdk" --add-modules javafx.web,javafx.fxml,javafx.swing,javafx.media --add-opens=javafx.fxml/javafx.fxml=ALL-UNNAMED -jar /app/scripture-app-builder/bin/scripture-app-builder.jar "$@"
```

chrisvire replied:
This should not be quoted since it could be a series of command-line arguments.

hfiguiere replied:
no. 

```
@      Expands to the positional parameters, starting from one. [...] When the expansion occurs within double quotes, each parameter expands to a separate word.  That
is, "$@" is equivalent to "$1" "$2" ...
```

chrisvire replied:
Thanks for the clarification.  I will try that out.

chrisvire replied:
Changed.  This worked.  Thanks!

hfiguiere commented on xterm.yml:
this should be in the main manifest since it's not autogenerated.

hfiguiere replied:
but then I question building `xterm`

chrisvire replied:
Yeah, I don't like it. One of the reasons for creating a Flatpak is so that we can work on multiple distributions/desktop flavors.  One of the output types from our platform is a PWA.  To test it, the user need to run `vite preview` which launches hosts a web server. We want that in a terminal window so that it is obvious that it is a separate running process, they can interact with the vite process, and close it when they are done.

When we were creating a Debian package for Ubuntu, we would just launch the gnome terminal.  That doesn't seem appropriate if we want to run on a wide variety of dsektops like Gnome, KDE, XFCE, etc. 

Do you have a better suggestion?

chrisvire replied:
> this should be in the main manifest since it's not autogenerated.

It seemed better to modularize it.

hfiguiere replied:
no. it's not modularize. its just complexified. It's hard to maintain.

chrisvire replied:
I have combined xterm.xml and openjdk.xml into the main manifest.

hfiguiere commented on xterm.yml:
```diff
@@ -0,0 +1,40 @@
+        # OLD
+        # url: https://www.x.org/archive/individual/lib/libXaw-1.0.15.tar.xz
+        # sha256: ab35f70fde9fb02cc71b342f654846a74328b74cb3a0703c02d20eddb212754a
+  - name: libXaw-32bit
```
what is this for?

chrisvire replied:
I will see if this can be removed.

chrisvire replied:
Removed libXaw-32bit.

chrisvire commented:
> How large are these and are they related to one another ? If size is not an issue having them in one Flatpak is better similar to how Libreoffice works.

We proposed that to the lead developer.  For "political" reasons, they should be separate.

chrisvire commented:
@bbhtt @hfiguiere I believe I have made all the requested changes and hope I have addressed all of your concerns.

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,86 @@
+    - type: file
+      only-arches:
+      - x86_64
+      url: https://sil-app-builders-linux-artifacts.s3.amazonaws.com/artifacts/59/scripture-app-builder.tgz
+      sha256: 1689bf31291802614401042f5c263c1c30d10a546982570ada7ac2d6924e77ff 
```
This is unversioned ? It will break whenever the source is rotated because sha256 will not match anymore.

chrisvire replied:
It is versioned by the build number (59).  Every time we do a release, I will update the url and the sha256 code.

bbhtt commented on python3-aeneas.yml:
```diff
@@ -0,0 +1,49 @@
+      sha256: b80878c9f40111313e55da8ba20bdba06d8fa3969fc68304167741bbf9e082ed
+modules:
+    - name: espeak
+      # buildsystem: simple
+      # build-commands:
+      #     - |
+      #         set -xueo pipefail
+      #         cd src
+      #         make install
```
```suggestion
```

chrisvire replied:
Sorry!  I commented it out since I wasn't sure it would work. and forgot to delete it.

chrisvire replied:
The whole file has been removed and the contents moved to the main manifest.  It does not contain the above commented out code.

chrisvire commented:
@bbhtt I think it is ready this time.

chrisvire commented:
@bbhtt @hfiguiere I did another pass through the files and cleaned up some comments.  Thanks for reviewing!

chrisvire commented:
@bbhtt @hfiguiere Please let me know what is left to be addressed.  Thanks.

chrisvire commented:
@bbhtt @hfiguiere I have removed our download of node20 and using the node20 extension.

What else do I need to do?

Thanks!

chrisvire commented:
@bbhtt @hfiguiere, I have:
* combined all non-autogen manifest entries into the same file
* removed access to /tmp
* changed to use the node20 extension

What else would you like changed?

bbhtt commented on org.sil.scripture-app-builder.yml:
```diff
@@ -0,0 +1,160 @@
+      - mkdir /app/jdk17.0.11
+      - mv /app/jdk/* /app/jdk17.0.11/
+      - rm -r /app/jdk
+    finish-args:
+      - "--env=PATH=/app/jre/bin:/usr/bin:$PATH"
```
```suggestion
```

This does not work here. Also variable substitution does not work in `finis-args`

You need to pass the entire thing: `--env=PATH=/app/jre/bin:/app/bin:/usr/bin`

chrisvire replied:
I have cleaned up openjdk installation.

chrisvire commented:
@bbhtt @hfiguiere So that you know, the original work on creating the flatpak packaging of our software was an intern project.  The result worked and I didn't completely understand all the parts of flatpak packaging.  So I submitted to get feedback.

Your feedback was invaluable!  Over the past couple of days, I feel I have been able to clean up and reorganize it into a more easily manageable script.

The latest changes:
* Reorganize the modules into layers from least likely to change to most likely to change. It seems like flatpak is similar to docker and this would be better.
* Simplify the openjdk module.
* Break out the openjfx module and use paths that are more relevant (/app/lib/openjfx instead of /app/lib/sdk)
* Simplify the extraction of prebuilt binaries into the proper place from the beginning and install desktop entry and metainfo from that directory.
* Clean up files from extraction that are not needed in that directory.

Please let me know of any additional suggestions that you have.

bbhtt commented:
bot, build org.sil.scripture-app-builder

bbhtt commented:
Please test if it's working.

chrisvire commented:
@bbhtt The test build passes.  Thanks.

bbhtt commented:
Did you install it and run ?

chrisvire commented:
I installed and ran it and did the following tests:
- launched app from search (uses desktop settings)
- build Android app (uses JDK and top)
- build and run PWA app ( uses node20 and xterm)
- fine tune timings (uses resources)
- synchronize text/audio (uses python and Aeneas libraries)
- used book viewer (uses openjfx)

chrisvire commented:
@bbhtt: I reviewed the changes with our intern and we found some more simplifications:
1. the python3-requests.yaml file was only needed for the python scripts that extracted the docker file so those are not needed anymore.
2. the run shell script is only one line long so I saw (when looking that the node20 manifest) that it could be included in the manifest.

Do we need to do another test build?

bbhtt commented:
You can start them if you want

bbhtt commented:
bot, build org.sil.scripture-app-builder

chrisvire commented:
I tested the build and it was successful.  Thanks!

chrisvire commented:
@bbhtt Is there anything else to resolve? I am curious about the timetable for the next step.  Thanks.

bbhtt commented:
No from you, the process is explained in https://docs.flathub.org/docs/for-app-authors/submission#faq
